sharedbuffers/mapped_struct.py

```diff
@@ -3177,124 +3177,214 @@ def __set__(self, obj, value):
     def __delete__(self, obj):
         raise TypeError("Proxy objects are read-only")
 
+
+if cython.compiled:
+    # We need as many of these definitions as different parameters are
+    # used per-function; otherwise, cython will deduce that they are of
+    # the same type. Sigh ...
+
+    numeric_A = cython.fused_type(
+        cython.char,
+        cython.uchar,
+        cython.short,
+        cython.ushort,
+        cython.int,
+        cython.uint,
+        cython.long,
+        cython.ulong,
+        cython.longlong,
+        cython.ulonglong,
+        cython.float,
+        cython.double,
+    )
+
+    numeric_B = cython.fused_type(
+        cython.char,
+        cython.uchar,
+        cython.short,
+        cython.ushort,
+        cython.int,
+        cython.uint,
+        cython.long,
+        cython.ulong,
+        cython.longlong,
+        cython.ulonglong,
+        cython.float,
+        cython.double,
+    )
+else:
+    globals().update(dict(
+        numeric_A = object,
+        numeric_B = object,
+    ))
+
+
+if cython.compiled:
+    @cython.cfunc
+    @cython.locals(self = BaseBufferProxyProperty, elem = numeric_A, obj = BufferProxyObject)
+    def _c_buffer_proxy_get_gen(self, obj, elem):
+        if obj is None:
+            return self
+        elif obj.none_bitmap & self.mask:
+            return None
+        else:
+            assert (obj.offs + self.offs + cython.sizeof(elem)) <= obj.pybuf.len   #lint:ok
+            mfence_full()   # acquire
+            return cython.cast('numeric_A *',
+                cython.cast(cython.p_uchar, obj.pybuf.buf) + obj.offs + self.offs)[0]   #lint:ok
```

Comment:
I think we should compute the pointer before the fence, and only do

``` python
pvalue = ...
mfence_full()
return pvalue[0]
```

Otherwise, the compiler/cpu will have to re-read all proxy attributes from memory.

Reply:
Good idea.